### ISSUE TITLE
Hiccup

### ISSUE THREAD

**Issue by den1k** [hyperfiddle/electric#32]
[Here](https://github.com/roman01la/uix/blob/0da33eef38a7122be226b9b9a8ae0b5431b6b5d3/core/src/uix/compiler/aot.clj#L494-L498) hiccup is transformed into react elements statically at macroexpansion time and otherwise at runtime. For the runtime case it should be easy to swap react for electric-dom2 calls but I'm not sure how reactive updates would work.

What's the current thinking on challenges to support hiccup for DOM generation?

**Comment by den1k** [comment]
Hiccup might be a force of habit after all. The more I use electric's syntax the more I find that it is actually more powerful and less of a headache than hiccup. Hiccup might be great for serialization (which doesn't make much sense in electric's case anyway). The only thing I'm looking forward to is more brevity but all the above makes sense; the foundation needs to be robust.

Since a lot of people have written a considerable amount of code in electric's current syntax, maybe it would be worthwhile to consider a code-rewriting tool in the future? Could also be a community effort. The highest friction I've experienced with electric as an early adopter has been rewriting between `electric-domx` versions.

**Comment by den1k** [comment]
bonus: it could be used on all existing electric examples

**Comment by dustingetz** [comment]
List of hiccup integration attempts (or other non-hiccup sugar over electric):
* https://clojurians.slack.com/archives/C7Q9GSHFV/p1682811661028259
* https://github.com/hyperfiddle/electric/blob/master/src/contrib/element_syntax.cljc
* ... what others?

**Comment by dustingetz** [comment]
List of problem statements:

Mark:

> Ultimately, my motivating factor is cultural:  I have a team of Clojure newbies who are falling in love with Clojure’s data oriented style and that they can solve almost any problem with the proper combination of map , reduce, and filter.  I also need to build lots of UI quickly. 
> 
> Electric is a new language with a different set of biases than Clojure.  I began thinking about the pedology of my problem and realized I would spend a bunch of my time saying things like, “That doesn’t work here.  Remember, you’re in Electric-land not in Clojure-land.”  In an effort to reduce that problem, I wondered if hiccup would allow of a smoother on-ramp for developers moving from Python->Clojure->lots of UI.
> 
> Said another way, I was trying to explore the question, does hiccup reduce the seams between Clojure and Electric?
> 
> As a result of this exercise, I think I agree with @denik’s comment “Hiccup might be great for serialization (which doesn’t make much sense in electric’s case anyway)”

**Comment by dustingetz** [comment]
https://github.com/lynaghk/hicada

**Comment by dustingetz** [comment]
Quotes from people who changed their mind:

* Denik, above: Hiccup might be a force of habit after all. The more I use electric's syntax the more I find that it is actually more powerful and less of a headache than hiccup.
* telekid [[slack](https://clojurians.slack.com/archives/C7Q9GSHFV/p1689168702004869)]: FWIW after some initial grumbling, I'm not missing hiccup syntax as much as I thought I would
* jjttjj: I really like the electric dom syntax. I kind of even like that it's a little more verbose than things like hoplon for example, where in electric you must explicitly wrap props and text for example, because it makes sense to me that these should be explicit, since that is like, the granularity where things can change in your dom tree.

**Comment by eneroth** [comment]
FWIW, I don't have a preference for one over the other. For me, Hiccup is surface-level stuff, and semantics and performance matter more. But I've been through UIx and Fulcro as well, who have chosen to stay away from Hiccup because of performance and other reasons.

I remember Roman of UIx quoting that Hiccup parsing was a significant percentage of each rendering cycle at Pitch, before they moved to UIx.

**Comment by conjurernix** [comment]
My issue with the current dom API is not the "syntax" per say, I don't have a preference over hiccup or sexprs. It's the effectful evaluation that bothers me. Evaluating `(dom/p (dom/text "Hello"))` doesn't result to some object representing a dom node. It evaluates to nil, with the effect of creating a paragraph element and attaching it to the dome side-effectfully. Another example is `dom/props` is a side-effectful procedur that attaches the props to the 'current node' whatever that may be if you try to be more _dynamic_.

This leads to a whole lote unintuitive, unclojure-like API that I grew to like. This became very apparent to me when i tried creating reusable components and i wanted to have "higher-ordered" components, and the evaluation model caught me by surprise. https://clojurians.slack.com/archives/C7Q9GSHFV/p1715512802650349

I'm sure the more I use it the more used to it i'll get, but I'm also sure there'll be some other corner cases where the said effectfulness will bite me in the a.

**Comment by conjurernix** [comment]
> My issue with the current dom API is not the "syntax" per say, I don't have a preference over hiccup or sexprs. It's the effectful evaluation that bothers me. Evaluating `(dom/p (dom/text "Hello"))` doesn't result to some object representing a dom node. It evaluates to nil, with the effect of creating a paragraph element and attaching it to the dome side-effectfully. Another example is `dom/props` is a side-effectful procedur that attaches the props to the 'current node' whatever that may be if you try to be more _dynamic_.
> 
> This leads to a whole lote unintuitive, unclojure-like API that I grew to like. This became very apparent to me when i tried creating reusable components and i wanted to have "higher-ordered" components, and the evaluation model caught me by surprise. https://clojurians.slack.com/archives/C7Q9GSHFV/p1715512802650349
> 
> I'm sure the more I use it the more used to it i'll get, but I'm also sure there'll be some other corner cases where the said effectfulness will bite me in the a.

An update on this comment: https://clojurians.slack.com/archives/C7Q9GSHFV/p1716072530502729

TL;DR I grew quite fond of the current DOM API that `electric` provides.

**Comment by milelo** [comment]
The https://github.com/luchiniatwork/hiccup-for-electric proposal is the functionality I'd like to see - from my novice perspective. Any informed comments on this?

**Comment by den1k** [comment]
Closing as I opened the issue and came to the conclusion hiccup would be a limiting force in electric as described [here](https://github.com/hyperfiddle/electric/issues/32#issuecomment-1478393433)

**Comment by milelo** [comment]
I've created a library that provides hiccup-like syntactic sugar for [Electric V2](https://github.com/hyperfiddle/electric). This can be intermingled with regular electric syntax.

https://github.com/milelo/electric-hiccup.
